### PR TITLE
range-ify reader

### PR DESCRIPTION
**Pull request by Herringway**
Reader is now a range. Scanner still depends on a lot of non-range bits, though. Going to eliminate those.

This is very much a WIP.

### CODE REVIEW

File: source/dyaml/scanner.d [hunk 1]
```diff
@@ -740,8 +745,8 @@ struct Scanner
         {
             // Check one char first, then all 3, to prevent reading outside the buffer.
             return reader_.column     == 0     &&
-                   reader_.peekByte() == '.'   &&
-                   reader_.prefix(3)  == "..." &&
+                   reader_.front == '.'   &&
+                   reader_.slice(3)  == "..." &&
```
Comment (Basile-z) [inline source/dyaml/scanner.d]: how about supporting the true slicer operator, e.g: `reader_[0..3]` ?

File: source/dyaml/reader.d [hunk 2]
```diff
@@ -174,69 +174,22 @@ final class Reader
             return d;
         }
 
-        /// Optimized version of peek() for the case where peek index is 0.
-        dchar peek() @safe pure
+        bool empty() @safe pure
```
Comment (Basile-z) [inline source/dyaml/reader.d]: nothrow

File: source/dyaml/scanner.d (hunk 1 above)
Comment (Herringway) [reply source/dyaml/scanner.d]: true slices don't actually cover the same functionality. `.slice(x)` would translate to something more like reader_[curPos..curPos+x], but with decoding. My goal is to have it removed.

Comment (Basile-z) [reply source/dyaml/scanner.d]: okay, dismiss the comment then

File: source/dyaml/reader.d (hunk 2 above)
Comment (Basile-z) [reply source/dyaml/reader.d]: and const

**Comment by Herringway** [comment]
Next: eliminate SliceBuilder.

**Pull request closed without merging**